File: Array/Easy/448_Find_All_Numbers_Disappeared_In_An_Array/Solution.py
from typing import List


class Solution:
    def findDisappearedNumbers(self, nums: List[int]) -> List[int]:
        i = 0
        n = len(nums)
        
        while i < n:
            correct_index = nums[i] - 1
            if nums[i] != nums[correct_index]:
                # Swap the number to its correct position
                nums[i], nums[correct_index] = nums[correct_index], nums[i]
            else:
                i += 1  
        
        missing_numbers = []
        for i, num in enumerate(nums):
            if num != i + 1:
                missing_numbers.append(i + 1)

        return missing_numbers
                

File: Array/Easy/448_Find_All_Numbers_Disappeared_In_An_Array/Solution.cpp
#include <vector>


class Solution {
public:    
    std::vector<int> findDisappearedNumbers(std::vector<int>& nums) {
        int i{}, n = nums.size();
        while (i < n) {
            if (nums[i] != nums[nums[i] - 1]) {
                std::swap(nums[i], nums[nums[i] - 1]);
            }
            else {
                ++i;
            }
        }

        std::vector<int> res;
        for (int i{}; i < n; ++i) {
            if (nums[i] != i + 1) {
                res.push_back(i + 1);
            }
        }
        return res;
    }
};